primitives/src/vid/advz.rs

```diff
@@ -445,30 +493,48 @@ where
         }
 
         // all shares must have equal evals len
-        let num_polys = shares
+        let num_evals = shares
             .first()
             .ok_or_else(|| VidError::Argument("shares is empty".into()))?
             .evals
             .len();
         if let Some((index, share)) = shares
             .iter()
             .enumerate()
-            .find(|(_, s)| s.evals.len() != num_polys)
+            .find(|(_, s)| s.evals.len() != num_evals)
         {
             return Err(VidError::Argument(format!(
                 "shares do not have equal evals lengths: share {} len {}, share {} len {}",
                 0,
-                num_polys,
+                num_evals,
                 index,
                 share.evals.len()
             )));
         }
+        if num_evals % self.multiplicity > 0 {
+            return Err(VidError::Argument(format!(
+                "multiplicity does not divide num_evals: multiplicity {}, num_evals {}",
+                self.multiplicity, num_evals,
+            )));
+        }
+        let chunk_size = self.multiplicity * self.payload_chunk_size;
+        let num_polys = num_evals / self.multiplicity;
 
-        let elems_capacity = num_polys * self.payload_chunk_size;
+        let elems_capacity = num_evals * chunk_size;
         let mut elems = Vec::with_capacity(elems_capacity);
-        for i in 0..num_polys {
+
+        for p in 0..num_polys {
+            // each share is guaranteed to contain the same number of evals
+            // but we don't know to which polynomial each eval belongs
```

Comment:
As discussed offline, let's strengthen this check to ensure for all shares the evals len equals `common.poly_commits.len()` times `multiplicity`. Then you can know which polynomial belongs to each eval.